packages/ts/src/rules/explicitModuleBoundaryTypes.ts

```diff
@@ -0,0 +1,171 @@
+// eslint-disable perfectionist/sort-switch-case -- cases are ordered by importance, not alphabetically
+import type { ReportMessageData } from "@flint.fyi/core";
+import type { RuleContext } from "@flint.fyi/core/src/types/context.js";
+
+import * as ts from "typescript";
+
+import { getTSNodeRange } from "../getTSNodeRange.js";
+import { typescriptLanguage } from "../language.js";
+
+const messages = {
+	fnParam: {
+		primary:
+			"Public facing functions should have explicit types for all parameters.",
+		secondary: ["{{ parameterName }} is declared here."],
+		suggestions: [],
+	},
+	fnReturn: {
+		primary: "Public facing functions should have an explicit return type.",
+		secondary: [],
+		suggestions: [],
+	},
+	fnReturnWithSuggestion: {
+		primary: "Public facing functions should have an explicit return type.",
+		secondary: [],
+		suggestions: [
+			"Add `{{ returnType }}` to the return type of this function.",
+		],
+	},
+} satisfies Record<string, ReportMessageData>;
+
+export default typescriptLanguage.createRule({
+	about: {
+		id: "explicitModuleBoundaryTypes",
+		preset: "restrictive",
+	},
+	messages,
+	setup(ctx) {
+		const sourceFile = ctx.sourceFile;
+		// ignore .d.ts and .js files
+		if (sourceFile.isDeclarationFile || isJavaScriptFile(sourceFile)) {
+			return;
+		}
```

Comment:
Ideating: maybe this should enabled in a preset like `"isolatedDeclarations"` that's only meant to be enabled on TS source files? I've always found it confusing when rules enforce their own file restrictions separate from the user's config.

Plus, I think some users will still want this in JS files if they're typed with `checkJs` & JSDocs?